core-bundle/src/EventListener/UndoOperationsListener.php

```diff
@@ -0,0 +1,149 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener;
+
+use Contao\Backend;
+use Contao\Controller;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\CoreBundle\ServiceAnnotation\Callback;
+use Contao\Image;
+use Contao\StringUtil;
+use Doctrine\DBAL\Connection;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+class UndoOperationsListener
+{
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    /**
+     * @var ContaoFramework
+     */
+    private $framework;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var Controller
+     */
+    private $controller;
+
+    /**
+     * @var Backend
+     */
+    private $backend;
+
+    /**
+     * @var Image
+     */
+    private $image;
+
+    public function __construct(ContaoFramework $framework, Connection $connection, TranslatorInterface $translator)
+    {
+        $this->framework = $framework;
+        $this->connection = $connection;
+        $this->translator = $translator;
+
+        /** @var Controller $controller */
+        $controller = $framework->getAdapter(Controller::class);
+        $this->controller = $controller;
+
+        /** @var Backend $backend */
+        $backend = $framework->getAdapter(Backend::class);
+        $this->backend = $backend;
+
+        /** @var Image $image */
+        $image = $this->framework->getAdapter(Image::class);
+        $this->image = $image;
+    }
+
+    /**
+     * @Callback(table="tl_undo", target="list.operations.undo.button")
+     */
+    public function onRenderUndoButton(array $row, ?string $href, string $label, string $title, ?string $icon, string $attributes = ''): string
```

Comment:
since this listener has only one public method (as it should!), the method should be `__invoke` and you can place the `@Callback` annotation directly on the class. Also please mark the class as `@internal` please.

Reply:
I agree! Since this PR is related to #3498 I would then also add a separate listener for the `jumpToParent` there and move everything to a `EventListener/Undo`  sub-namespace?

Reply:
I implemented this in https://github.com/contao/contao/pull/3498/commits/2f8a4cdaba576237bbcd73fca5f9933fba282b93. If nobody objects to this I would update this PR here accordingly.